### Vanguard Project - Client Profiles df

This notebook focuses on the analysis of the df_final_demo dataset, which contains demographic and account information about Vanguard clients.

The main objectives are to understand the structure and content of the dataset, assess its data quality, identify potential inconsistencies, and determine whether any issues could affect the subsequent analysis.

In [2]:
import pandas as pd

In [28]:
url = "https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_demo.txt"
df_final_demo = pd.read_csv(url, sep=",")
df_final_demo.head()

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
0,836976,6.0,73.0,60.5,U,2.0,45105.30,6.0,9.0
1,2304905,7.0,94.0,58.0,U,2.0,110860.30,6.0,9.0
2,1439522,5.0,64.0,32.0,U,2.0,52467.79,6.0,9.0
3,1562045,16.0,198.0,49.0,M,2.0,67454.65,3.0,6.0
4,5126305,12.0,145.0,33.0,F,2.0,103671.75,0.0,3.0


In [29]:
df_final_demo.info()

<class 'pandas.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   client_id         70609 non-null  int64  
 1   clnt_tenure_yr    70595 non-null  float64
 2   clnt_tenure_mnth  70595 non-null  float64
 3   clnt_age          70594 non-null  float64
 4   gendr             70595 non-null  str    
 5   num_accts         70595 non-null  float64
 6   bal               70595 non-null  float64
 7   calls_6_mnth      70595 non-null  float64
 8   logons_6_mnth     70595 non-null  float64
dtypes: float64(7), int64(1), str(1)
memory usage: 4.8 MB


Most variables are numerical, while `gendr` is categorical. The `client_id` variable contains no missing values, whereas the remaining variables contain a small number of missing observations.

In [30]:
df_final_demo.isna().sum()

client_id            0
clnt_tenure_yr      14
clnt_tenure_mnth    14
clnt_age            15
gendr               14
num_accts           14
bal                 14
calls_6_mnth        14
logons_6_mnth       14
dtype: int64

In [31]:
df_final_demo[df_final_demo.isna().any(axis=1)]

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
4164,7402828,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8316,355337,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8677,8412164,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9583,4666211,8.0,106.0,NaN,F,2.0,42550.55,4.0,7.0
13444,2222915,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18066,4876926,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25961,5277910,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
28432,7616759,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
35323,8191345,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
43518,1227228,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


We identified 15 clients with missing information. For 14 clients, all profile variables are missing, while one client has only the age missing.

In [32]:
df_final_demo["client_id"].nunique() == df_final_demo.shape[0]

True

In [33]:
df_final_demo[["clnt_tenure_yr", "clnt_tenure_mnth"]].describe()

,clnt_tenure_yr,clnt_tenure_mnth
count,70595.000000,70595.000000
mean,12.052950,150.659367
std,6.871819,82.089854
min,2.000000,33.000000
25%,6.000000,82.000000
50%,11.000000,136.000000
75%,16.000000,192.000000
max,62.000000,749.000000


In [34]:
df_final_demo["tenure_diff"] = (df_final_demo["clnt_tenure_mnth"] - df_final_demo["clnt_tenure_yr"] * 12)
df_final_demo["tenure_diff"].describe()

count    70595.000000
mean         6.023968
std          3.814399
min          0.000000
25%          2.000000
50%          6.000000
75%         10.000000
max         12.000000
Name: tenure_diff, dtype: float64

In [35]:
df_final_demo[
    ["clnt_tenure_yr", "clnt_tenure_mnth", "tenure_diff"]
].head(20)

,clnt_tenure_yr,clnt_tenure_mnth,tenure_diff
0,6.0,73.0,1.0
1,7.0,94.0,10.0
2,5.0,64.0,4.0
3,16.0,198.0,6.0
4,12.0,145.0,1.0
5,5.0,71.0,11.0
6,5.0,66.0,6.0
7,30.0,361.0,1.0
8,30.0,369.0,9.0
9,15.0,189.0,9.0


The difference between total months and years converted to months ranges from 0 to 12 months.

This suggests that `clnt_tenure_yr` represents completed years of tenure, while `clnt_tenure_mnth` represents total tenure in months. 

Therefore, the apparent difference between the two variables does not indicate a data inconsistency. The tenure variables appear to be consistent.

In [36]:
df_final_demo.drop(columns="tenure_diff", inplace=True)

In [37]:
profile_columns = ["clnt_tenure_yr", "clnt_tenure_mnth", "clnt_age", "gendr", "num_accts", "bal", "calls_6_mnth", "logons_6_mnth"]
missing_profile = df_final_demo[df_final_demo[profile_columns].isna().all(axis=1)]
missing_profile

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
4164,7402828,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8316,355337,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8677,8412164,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
13444,2222915,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18066,4876926,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25961,5277910,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
28432,7616759,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
35323,8191345,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
43518,1227228,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
46076,8611797,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [26]:
missing_profile["client_id"]

4164     7402828
8316      355337
8677     8412164
13444    2222915
18066    4876926
25961    5277910
28432    7616759
35323    8191345
43518    1227228
46076    8611797
47189    5144725
49846    1037867
63701    1618749
66219    6021001
Name: client_id, dtype: int64

#### Investigation missing profile information

In [39]:
url_2 = "https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_experiment_clients.txt"
df_final_experiment_clients = pd.read_csv(url_2, sep=",")
df_final_experiment_clients.head()

,client_id,Variation
0,9988021,Test
1,8320017,Test
2,4033851,Control
3,1982004,Test
4,9294070,Control


In [44]:
df_final_experiment_clients.shape

(70609, 2)

In [40]:
missing_profile_experiment = df_final_experiment_clients[
    df_final_experiment_clients["client_id"].isin(missing_profile["client_id"])
]

missing_profile_experiment

,client_id,Variation
5171,8191345,Control
7131,5144725,Test
8902,7616759,Control
18742,2222915,Test
19391,5277910,Test
23678,8412164,Test
36137,8611797,Control
36245,355337,Control
38599,1227228,Test
48110,1037867,Test


All 14 clients with completely missing profile information are present in the experiment roster.

Among these clients, 6 are assigned to the Control group, 6 to the Test group, and 2 have no experimental variation recorded.

In [42]:
df_final_experiment_clients["Variation"].value_counts(dropna=False)

Variation
Test       26968
Control    23532
NaN        20109
Name: count, dtype: int64

The `client_id` is consistent across the two datasets: all 70,609 clients in `df_final_demo` are also present in `df_final_experiment_clients`.

However, the experimental assignment is not complete. While 26,968 clients are assigned to the Test group and 23,532 to the Control group, 20,109 clients have no recorded value for `variation`.

In [47]:
clients_no_variation = df_final_experiment_clients[df_final_experiment_clients["Variation"].isna()]
clients_no_variation

,client_id,Variation
50500,5459747,NaN
50501,8031000,NaN
50502,1847030,NaN
50503,9713157,NaN
50504,7775828,NaN
...,...,...
70604,2443347,NaN
70605,8788427,NaN
70606,266828,NaN
70607,1266421,NaN


In [51]:
demo_no_variation = df_final_demo[df_final_demo['client_id'].isin(clients_no_variation['client_id'])]
demo_no_variation.describe()

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,num_accts,bal,calls_6_mnth,logons_6_mnth
count,2.010900e+04,20107.000000,20107.000000,20107.000000,20107.000000,2.010700e+04,20107.000000,20107.000000
mean,5.002010e+06,12.106232,151.271746,44.238773,2.257920,1.422490e+05,4.108619,4.147710
std,2.876933e+06,6.900585,82.451745,15.555909,0.538317,3.001242e+05,2.192047,2.178329
min,1.690000e+02,2.000000,33.000000,13.500000,1.000000,1.378942e+04,1.000000,1.000000
25%,2.531946e+06,6.000000,82.000000,30.500000,2.000000,3.014772e+04,2.000000,2.000000
50%,4.998573e+06,11.000000,138.000000,45.000000,2.000000,5.633810e+04,4.000000,4.000000
75%,7.493150e+06,16.000000,193.000000,56.500000,2.000000,1.305127e+05,6.000000,6.000000
max,9.999839e+06,62.000000,749.000000,92.500000,8.000000,1.283842e+07,7.000000,7.000000
